plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/experimental/ViewportPluginImpl.kt

```diff
@@ -0,0 +1,210 @@
+package com.mapbox.maps.plugin.viewport.experimental
+
+import com.mapbox.android.gestures.MoveGestureDetector
+import com.mapbox.maps.plugin.animation.Cancelable
+import com.mapbox.maps.plugin.delegates.MapDelegateProvider
+import com.mapbox.maps.plugin.gestures.GesturesPlugin
+import com.mapbox.maps.plugin.gestures.OnMoveListener
+import com.mapbox.maps.plugin.gestures.gestures
+import com.mapbox.maps.plugin.viewport.VIEWPORT_STATUS_OBSERVER_REASON_PROGRAMMATIC
+import com.mapbox.maps.plugin.viewport.VIEWPORT_STATUS_OBSERVER_REASON_USER_INTERACTION
+import com.mapbox.maps.plugin.viewport.experimental.data.DefaultViewportTransitionOptions
+import com.mapbox.maps.plugin.viewport.experimental.data.FollowingViewportStateOptions
+import com.mapbox.maps.plugin.viewport.experimental.data.OverviewViewportStateOptions
+import com.mapbox.maps.plugin.viewport.experimental.state.FollowingViewportState
+import com.mapbox.maps.plugin.viewport.experimental.state.FollowingViewportStateImpl
+import com.mapbox.maps.plugin.viewport.experimental.state.OverviewViewportState
+import com.mapbox.maps.plugin.viewport.experimental.state.OverviewViewportStateImpl
+import com.mapbox.maps.plugin.viewport.experimental.state.ViewportState
+import com.mapbox.maps.plugin.viewport.experimental.transition.DefaultViewportTransition
+import com.mapbox.maps.plugin.viewport.experimental.transition.DefaultViewportTransitionImpl
+import com.mapbox.maps.plugin.viewport.experimental.transition.ImmediateViewportTransition
+import com.mapbox.maps.plugin.viewport.experimental.transition.ViewportTransition
+import java.lang.IllegalStateException
+import java.util.concurrent.CopyOnWriteArrayList
+import java.util.concurrent.CopyOnWriteArraySet
+
+/**
+ * Mapbox default implementation for [ViewportPlugin].
+ */
+class ViewportPluginImpl : ViewportPlugin {
+  private val registeredStatusObservers = CopyOnWriteArrayList<ViewportStatusObserver>()
+  private val internalStates = CopyOnWriteArraySet<ViewportState>()
+  private var currentCancelable: Cancelable? = null
+  private val transitions = HashMap<Pair<ViewportState?, ViewportState?>, ViewportTransition>()
+  private lateinit var delegateProvider: MapDelegateProvider
+  private lateinit var gesturesPlugin: GesturesPlugin
+  private val onMoveListener = object : OnMoveListener {
+    override fun onMoveBegin(detector: MoveGestureDetector) {
+      if (transitionsToIdleUponUserInteraction) {
+        currentCancelable?.cancel()
+        currentCancelable = null
+        updateStatus(ViewportStatus.State(null), VIEWPORT_STATUS_OBSERVER_REASON_USER_INTERACTION)
+      }
+    }
```

Comment:
This whole logic would have to be customizable so that Nav SDK can introduce a more complex gesture handlers like https://github.com/mapbox/mapbox-navigation-android/blob/main/libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/camera/lifecycle/NavigationScaleGestureHandler.kt.

Reply:
The idea @macdrevx and I had is that we make the API flexible enough that users can implement their own gestures logic with current API(by updating the StateOptions on the fly). We provide very basic convenient methods to interrupt the ViewportState or ViewportTransition with user input.

I will take a look at the link and see if there's more we could do to help handling customised gestures.